README
		nspark - New Spark unarchiver
		=============================

THIS PROGRAM IS *FREE* -- YOU MAY DO WHAT YOU LIKE WITH IT AS LONG AS
THIS FILE REMAINS INTACT, AND YOU DO NOT MAKE ANY MONEY FROM IT!  IF
YOU MAKE IMPROVEMENTS TO THIS PROGRAM THEN PLEASE SEND A COPY OF THE
SOURCES TO THE MAINTAINER OF THAT VERSION OF THE PROGRAM.

nspark is a rewritten version of !Spark for UNIX. The original version
of spark (sometimes known as "cark") was based upon the BASIC program
"bark", written by David Pilling, and was improved and fixed by the
following people:

	Philip Colmer (pcolmer@acorn.co.uk),
	Alun Jones (auj@uk.ac.aber),
	Andy Duplain (duplain@rtf.bt.co.uk),
	Martin Percival (martin@thed.uk22.bull.com),
	Denis Howe (dbh@doc.ic.ac.uk).

I decided to re-write the program to add functionality and make it
easier to port to other Operating Systems (such as MSDOS and even RISC
OS :-).  The new version has the following enhancements:

	o Supports all archive methods used by !Spark 1 and 2. The old
	version didn't support "Squashing" and used to handle "Compressed"
	archives by using the UNIX uncompress program. The new version
	supports both these and has them built-in, allowing easier porting
	to other OS's.

	o Supports Cyclic Redundancy Checking (CRC). The integrity of an
	archive can now be checked without having to unarchive it.

 	o Supports verbose option to allow archive file details to be
 	displayed (time/date or exec/load address, size, type).

	o Supports selective file unarchiving, so the whole archive needn't
	be unarchived if not all the files are wanted).

	o Supports user-prompting before overwriting a file, allowing the
	user to either (1) overwrite the file, (2) overwrite all the
	files, (3) unarchive the file to a different name, or (4) not to
	unarchive the file at all.

	o Supports file-stamping so the date and time of the file, when
	unarchived, matches the date and time it was archived (not 4.2BSD
	or SysV.2 as these systems lack vital library functions).
			
I have not included support for creating !Spark archives, and I am not
really sure if I need to as: (1) !Spark for RISC OS is cheap (#5.99
from David Pilling), (2) !SparkPlug can unpack archives created with
zip (which is available in source form for UNIX, VMS and MSDOS), and
(3) it's a lot of work. Let me know what you think.

To compile nspark; copy the relevant makefile from the "mkfiles"
directory, and rename it "Makefile" or "makefile", edit it to suite C
compiler preference, change the "SYSTYPE" variable setting, etc, then
simply type "make" (or "nmake" or "amu" or whatever).

There are some problems with nspark if it is compiled with gcc v2 (it
core dumps).  To fix this try recompiling everything ("make clean"
first) with "-O" off, and if nspark still won't work try adding
"-fno-builtin" after "gcc" in the "CC" line, then recompile. I have
had some strange problems with optimisation in gcc v2, for example
basename() in misc.c core dumps (looking at the assembler gcc v2
produced I can see why!).
			
Please send bug reports/suggestions etc to the maintainer of the
particular version of nspark you are using (the maintainer's e-mail
address can be found using the -V option).

Enjoy.

--
Andy Duplain, BT Customer Systems, Brighton, UK.  (duplain@btcs.bt.co.uk).
October 1992.


CHANGES

1.2.4	7th Oct 92	1st public release.

1.2.5	9th Oct 92	Fixed bug in misc.c::riscos_path() that caused nspark
			to core dump under SCO UNIX.

1.2.6	19th Oct 92	Added -x option as an alternative to -u.  Changed
			unix.c::filestamp() so that timelocal(3) is only used
			for SunOS, and mktime(3) is used for all other
			flavours of UNIX.

1.2.7	23rd Oct 92	Minor cosmetic changes... I should really leave this
			thing alone.

1.2.8	6th Nov 92	Fixed function definition for unarc.c::prompt_user()
			(was defined the ANSI C way).

			Fixed io.c::read_header() so it doesn't read the
			load/exec/attr fields if the archive isn't an
			Archimedes archive.
			
			Corrected comments in compress.c about what crunch
			and squash methods were.
			
			Slight change to algorithm in crc.c::calccrc(), though
			no functional changes made.

1.2.9	9th Nov 92	compress.c: Moved offset and size from getcode() into
			global scope so that uncompress() can re-initialise
			them each time it is called.

			pack.c/pack.h: Added putc_init() to initialise the
			static variables used by putc_ncr().

1.2.10	12th Nov 92	Fixed bug with realloc() size in misc.c: uplevel().
			(reported by arb@comp.lancs.ac.uk).

1.2.11	7th Dec 92	Reformatted all source files using Emacs C-Mode (4-
			space tabs etc.).  Slight change to unarc.c::unarc()
			logic when printing file details during unarchiving.

1.3Beta 8th Dec 92	Added -a option to allow filetypes to be appended
			to filenames as they are unarchived.  This is to allow
			Acorn <-> UNIX NFS to work correctly.
			(requested by dhmyrdal@pvv.unit.no)

1.3	9th Dec 92	Changed -a option to -T, made filetypes append in
			lowercase hex rather than uppercase.

1.3.1	9th Dec 92	Changed ret from int in io.c::check_stream().  #ifdef'd
			out write_halfword() and write_word() in io.c/io.h .

1.3.2	22nd Dec 92	misc.c: Changed #include <malloc.h> to
			#include <stdlib.h>

1.4	23rd Dec 92	unarc.c: added total-printing if in verbose mode.

NEW MAINTAINER: Andrew Brooks, arb@comp.lancs.ac.uk.

1.5	5th Mar 93	Added code to compile on RISC OS: mkfiles/acorn,
			acorn.c, changes to io.c, unarc.c, main.c, io.h to
			add prototypes needed.

1.6	20th Aug 93     main.c: Add -C option to convert filenames to lowercase
			io.c: allow top-bit-set characters in RISCOS filenames.
			compress.c, unarc.c: prevent output if in quiet mode.

1.7	20th Aug 93	Added code to read headers from ArcFS archives.
			Changes to io.c, unarc.c, mkfiles.  Added arcfs.[ch]
			WARNING: Code is unfinished: only the list option (-l)
			can be used (and timestamps will be incorrect).  Trying
			to test or extract from the archive will fail, and
			probably dump core.  Please add the necessary
			code if you can, I don't have time at the moment.

1.7.1   4th Mar 94	Fixed many bugs in ArcFS code: changes to unarc.c,
			compress.c also.  Still no timestamps, and files
			extracted are too long by a few null bytes for some
			reason (complen is too large, but why?), so CRC fails.

1.7.2   26th Oct 94	Fixed date and time conversion for ArcFS archives.
			Added manual pages supplied by Duncan (?)
			(m91dps@ecs.oxford.ac.uk).

1.7.2a	9th Nov 94	Altered unix makefile to call ./mkendian not mkendian
			as suggested by Tom Tromey (tromey@busco.lanl.gov).
			Added rename scripts for RISC OS/UNIX.

1.7.3	12th Dec 94	Finally fix ArcFS archive extraction, fixes in
			arcfs.c, unarc.c, io.c and crc.c.  Modifications
			to manual pages to correct errors and add new details.

1.7.4	6th Jan 95	Added Windows NT and Alpha support by Alun Jones
			(auj@aber.ac.uk).
			Restructured directories and added executables for
			RISC OS, Sun OS 4.1.3 and Solaris 2.3.
			Added executables for Linux, MIPS and Alpha supplied
			by Alun Jones.
			Updated manual pages for new file formats.
			Added text versions of manual pages for those without
			nroff.

1.7.5	25th Jan 95	Bug fixes in compress.c, thanks to all who reported
			the problem.  Included binaries that report version
			1.7.4 should not be exhibit the bug as it was caused
			by a last-minute source change (mae culpa!).

1.7.6   01th Aug 95     DOS port. (Beta by Bob Brand, bob@wop.wtb.tue.nl,
			or B.A.Brand@wtb.tue.nl, addresses valid until
			at least Oct 1st, 1996), fido: 2:2801/208.2.
                        Removed floating point operation in
			arcfs.c::arcfs_read_header().
			Made htab[] and codetab[] arrays in compress.c
			dynamically allocated arrays (DOS version only!).
			Made misc.c::basename() insensitive for null-
			pointers. (IRIX's strlen() and a few other string-
			functions coredump on null-pointers. Should not
			happen under normal execution. But it can happen
			in the DOS version running under DOS 2.x.)
			Renamed io.h to sparkio.h (conflict with Borland
			C/C++ io.h).
			Removed `Does not support DOS' from BUGS file.
			Added `Does not time/date stamp directories
			under DOS or WINNT' to BUGS file.
			Included executables for SGI IRIX 4.x and 5.x.
			
1.7.7   17th March 1999 -I option Mark Usher, marku@magnet.at
		The -I option has been added to extract the individual
		files with an extra .inf file that contains the
		information relevant to that file. This should help
		people who use BBC emulators to read the files into
		disc images.
		The WinNT directory has been changed to Win32.



main.c

/*
 * main function
 *
 * Revision 1.12 99/03/17 MU
 * added the option -i to extract with .inf files
 * Also added the .inf extension to the default logfile name for MSDOS
 *
 * $Header: main.c 1.11 95/08/01 $
 * $Log:	main.c,v $
 * Revision 1.11  95/08/01  xx:xx:xx  BB
 * Fixed for Borland C/C++
 * Added / as command line switch option for DOS
 *
 * Revision 1.10  93/08/20  12:39:28  arb
 * Added support for ArcFS archive detection
 *
 * Revision 1.9  93/08/20  10:30:50  arb
 * Added -C option for "convert filenames to lowercase"
 *
 * Revision 1.8  93/03/05  15:40:32  arb
 * Added <stdlib.h> for RISCOS, needed for exit()
 *
 * Revision 1.7  92/12/09  09:43:03  duplain
 * Changed "-a" option to "-T".
 *
 * Revision 1.6  92/12/08  10:19:30  duplain
 * Added -a option for "append filetype".
 *
 * Revision 1.5  92/12/07  17:18:42  duplain
 * reformatted source.
 *
 * Revision 1.4  92/10/19  09:33:09  duplain
 * Added -x as an alternative to -u.
 *
 * Revision 1.3  92/10/01  11:21:39  duplain
 * Added -R option.
 *
 * Revision 1.2  92/09/30  10:27:29  duplain
 * Added logfile option and processing.
 *
 * Revision 1.1  92/09/29  18:02:20  duplain
 * Initial revision
 *
 */  
	
#include <stdio.h>
#include <ctype.h>
#include "spark.h"

/* BB changed next line */ 

/* #include "io.h" */ 
#include "nsparkio.h"
#include "error.h"
#include "misc.h"
#include "arc.h"
#include "unarc.h"
#include "garble.h"

/* BB changed next line */ 

/* #ifdef RISCOS */ 
#if defined(RISCOS) || defined(__MSDOS__)
#include <stdlib.h>
#endif	/* RISCOS || __MSDOS__ */

/* BB added next includes */ 
#ifdef __MSDOS__
#include <ctype.h>
#include <string.h>
#endif	/* __MSDOS__ */
	
#ifdef UNIX
static char rcsid[] = "$Header: main.c 1.11 95/08/01 $";

#endif	/* UNIX */
char *ourname;					/* program name */
char *archive;					/* name of archive file */


/* MU changed the default log filename for MSDOS */ 
#ifdef __MSDOS__
char *logfile = "settypes.inf";	/* default name for log file */

#else	/*  */
char *logfile = "settypes";		/* default name for log file */

#endif	/*  */
char **files;					/* optional file arguments */
unsigned char unarc = 0;		/* -u or -x or I */
unsigned char inffiles = 0;	/* I */
unsigned char quiet = 0;		/* -q */
unsigned char verbose = 0;		/* -v */
unsigned char testing = 0;		/* -t */
unsigned char listing = 0;		/* -l */
unsigned char force = 0;		/* -f */
unsigned char stamp = 1;		/* -s */
unsigned char retry = 0;		/* -R */
unsigned char apptype = 0;		/* -T */
unsigned char singlecase = 0;	/* -C */

#ifdef DEBUGGING
unsigned char debugging = 0;	/* -D */

#endif	/* DEBUGGING */
void usage P__((void));
int do_unarc P__((void));
int do_arc P__((void));

int
main(argc, argv)
	int argc;
	char **argv;
{
	register i;
	
		/* BB extra switch indicator for command line parsing
		   DOS allows switches like /a/b/c which enters as one
		   `word' in argv. */ 
#ifdef __MSDOS__
	int nextmaybeaswitch;
	
#endif	/*  */
		
#ifdef DEBUGGING
		/*
		 * check types are defined correctly for this machine (or compiler ???)
		 */ 
		if (sizeof(Word) != 4)
	{
		puts("Word size != 4");
		exit(1);
	}
	if (sizeof(Halfword) != 2)
	{
		puts("Halfword size != 2");
		exit(1);
	}
	if (sizeof(Byte) != 1)
	{
		puts("Byte size != 1");
		exit(1);
	}
	
#endif	/* DEBUGGING */
		ourname = basename(*argv++);
	
		/* BB cosmetics for DOS: Strip extention .exe (or .com if
		   somebody would like to compile nspark to a .com) from
		   the ourname string. And convert it to lowercase. That
		   way it looks better than ``NSPARK.EXE: error ...''. */ 
#ifdef __MSDOS__
		/* For DOS prior to 3.0, argv[0] contains the NULL pointer.
		   So substitute a `default'. */ 
		if (!ourname)
		ourname = "nspark";
	
		/* NB: stricmp == strcmpi == strcasecmp */ 
		if (stricmp(&ourname[strlen(ourname) - 4], ".com") == 0
			|| stricmp(&ourname[strlen(ourname) - 4], ".exe") == 0)
	{
		
			/* We cannot write a '\0' into ourname because it points to
			   argv[0]. And that may be in a read-only data segment. */ 
		char *newname, *cp;
		if ((newname = (char *) malloc(strlen(ourname)) - 3) != NULL)
		{
			strncpy(newname, ourname, strlen(ourname) - 4);
			newname[strlen(ourname) - 4] = '\0';
			ourname = newname;	/* Allocated space will be released
								   automatically at exit */
			for (cp = ourname; *cp; cp++)
				if (isascii(*cp) && isupper(*cp))
					*cp = tolower(*cp);
		}
	}
	
#endif	/* __MSDOS__ */
		argc--;
	
		/*
		 * parse args (can't use getopt() 'cos not all C libraries have it)
		 */ 
		while (argc)
	{
		int donext = 0;
		char *arg = *argv;
		
			/* BB changed next line */ 
			/* if (*arg == '-') { */ 
#ifdef __MSDOS__
			if (*arg == '-' || *arg == '/')
		{
			
#else	/*  */
			if (*arg == '-')
		{
			
#endif	/* __MSDOS__ */
			char c;
			
#ifdef __MSDOS__
				/* BB first char following a switch may not be another switch */ 
				nextmaybeaswitch = 0;
			
#endif	/* __MSDOS__ */
				while (!donext && !isspace(c = *++arg) && c)
			{
				switch (c)
				{
				case 'u':
				case 'x':
					unarc = 1;
					break;
				case 't':
					testing++;
					unarc = 1;	/* implied */
					break;
				case 'l':
					listing++;
					unarc = 1;	/* implied */
					break;
				case 'q':
					quiet = 1;
					break;
				case 'v':
					verbose = 1;
					break;
				case 'f':
					force = 1;
					break;
				case 's':
					stamp = 0;
					break;
				case 'R':
					retry = 1;
					break;
				case 'V':
					printf("%s v%s - maintained by %s - PUBLIC DOMAIN\n",
							ourname, VERSION, MAINTAINER);
					break;
				case 'T':
					apptype = 1;
					break;
				case 'C':
					singlecase = 1;
					break;
				case 'L':
					if (*++arg)
						logfile = arg;
					
					else
					if (--argc)
						logfile = *++argv;
					
					else
						usage();
					donext++;
					break;
					
#ifdef DEBUGGING
				case 'D':
					debugging = 1;
					break;
					
#endif	/* DEBUGGING */
#ifdef __MSDOS__
						/* BB DOS allows switches like /a/b/c */ 
				case '/':
					if (nextmaybeaswitch && arg[1] != '/')
						break;
					
					else		/* fall through to error message */
						;
					
#endif	/* __MSDOS__ */
				case 'I':
					unarc = 1;
					inffiles = 1;
					break;
				case 'p':
					if (*++arg)
						logfile = arg;
					else
						if (--argc)
							set_password(*++argv);
						else
							usage();
					donext++;
					break;
				default:
					error("unknown option '%c'", c);
					exit(1);
				}
				
#ifdef __MSDOS__
					/* BB We've had a valid switch, next char may be
					   a / again */ 
					nextmaybeaswitch = 1;
				
#endif	/* __MSDOS__ */
			}
			argv++;
			argc--;
		}
		else
			break;
	}
	if (!argc)
		usage();
	archive = *argv++;
	files = argv;
	if (unarc)
		i = do_unarc();
	
	else
		i = do_arc();
	exit(i);
	
		/* BB added next line */ 
		return 0;				/* Keep compiler happy. */
}



/*
 * display program usage and exit
 */ 
void
usage() 
{
	fprintf(stderr, "usage: %s [options] archive [file ... file]\n",
			 ourname);
	fprintf(stderr, "       where options are:\n");
	fprintf(stderr,
			 "       -u or -x unarchive           -t test archive integrity\n");
	fprintf(stderr, "       -l list archive contents     -q quiet\n");
	fprintf(stderr,
			 "       -f force file overwrite      -s no filestamp\n");
	fprintf(stderr,
			 "       -v verbose                   -V display version number\n");
	fprintf(stderr,
			 "       -R retry if archive corrupt  -L<name> set logfile name\n");
	fprintf(stderr,
			 "       -T append filetype to name   -C create lowercase filenames\n");
	
		/* MU added instuctions for the -I option */ 
	fprintf(stderr, "       -I unarchive with .inf files -p<password> set password\n");
	exit(1);
} 
